### PR TITLE
Add evaluation functionality for App Insights data

### PR DESCRIPTION
**Pull request by Abdulk084**
This PR introduces a new evaluation system for evaluating data from Azure Application Insights. The main features include:

- Implementation of `eval_combined.py`
- Extraction of multiple interactions from App Insights using `extract_app_insights_data`
- Support for three types of evaluations: groundedness, coherence, and relevance
- Flexible configuration to enable/disable specific evaluation types
- Processing of multiple entries 
- JSON output containing evaluation results for all processed entries

Key components:
- `evaluation_run`: Main function orchestrating the evaluation process
- `groundedness_evaluation`, `coherence_evaluation`, `relevance_evaluation`: Individual evaluation functions
- Error handling and logging for robust operation

Example output:
```[{"id": "chatcmpl-A8ibwkWtlnlbx4z5H17eGDeKQn0Jh", "groundedness_result": "3", "coherence_result": "4", "relevance_result": "4"}, {"id": "chatcmpl-A8ibaHVRUWTajpnIibekjhKio0CZt", "groundedness_result": "3", "coherence_result": "1", "relevance_result": "1"}, {"id": "chatcmpl-A8iWR5uSebHv21haStM9PqOCbR9E6", "groundedness_result": "3", "coherence_result": "stars: 1", "relevance_result": "1"}, {"id": "chatcmpl-A8iVsAZ5UuMcC2c1NtrRU3il3MCBl", "groundedness_result": "3", "coherence_result": "5", "relevance_result": "5"}, {"id": "chatcmpl-A8XOVgYFMM8mNONyTVDxT7ALu1e0s", "groundedness_result": "3", "coherence_result": "4", "relevance_result": "4"}]```

### CODE REVIEW

File: src/api/evaluators/custom_evals/eval_combined.py [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+logs_client = LogsQueryClient(credential)
+workspace_id= os.environ["APPINSIGHTS_WORKSPACE_ID"]
+
+
+def extract_app_insights_data( timespan_days: int = 2, take_count: int = 5) -> List[Dict[str, Any]]:
+    """
+    Extracts data from Azure Application Insights for multiple interactions.
+
+    Args:
+        workspace_id (str): The workspace ID to query.
```
Comment (drewby) [inline src/api/evaluators/custom_evals/eval_combined.py]: This currently a global, not an argument. It should probably be an argument, but I don't know all the rules of good python code. :)

File: src/api/evaluators/custom_evals/eval_combined.py [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+        timespan_days (int, optional): Number of days for the query timespan. Defaults to 2.
+        take_count (int, optional): Number of interactions to retrieve. Defaults to 5.
+
+    Returns:
+        List[Dict[str, Any]]: A list of dictionaries containing the extracted data.
+    """
+    try:
+        # Authenticate using DefaultAzureCredential
+        credential = DefaultAzureCredential()
```
Comment (drewby) [inline src/api/evaluators/custom_evals/eval_combined.py]: What's the difference between the credential here and the one on line 30?

File: src/api/evaluators/custom_evals/eval_combined.py [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
+
+
+# Load environment variables
+load_dotenv()
+
+
+
+# Configuration for the Azure OpenAI endpoint
+model_config = {
```
Comment (drewby) [inline src/api/evaluators/custom_evals/eval_combined.py:33]: check these are defined or exit with a message that they are required

File: src/api/evaluators/custom_evals/eval_combined.py [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+}
+
+
+# Authenticate using DefaultAzureCredential
+credential = DefaultAzureCredential()
+logs_client = LogsQueryClient(credential)
+workspace_id= os.environ["APPINSIGHTS_WORKSPACE_ID"]
+
+
+def extract_app_insights_data( timespan_days: int = 2, take_count: int = 5) -> List[Dict[str, Any]]:
```
Comment (drewby) [inline src/api/evaluators/custom_evals/eval_combined.py]: I think we should spend some time to refactor this into some shorter functions that fit on a screen for good readability. 

For example, create a function for get_interactions, get_responses, get_traces, etc. These could further reuse some logic in order to execute the query, get table[0], etc and return it, reducing duplicate code.

**Review by drewby — CHANGES_REQUESTED** [review]

**Review by sudivate — COMMENTED** [review]
[{"gen_ai.response.id": "chatcmpl-A8ibwkWtlnlbx4z5H17eGDeKQn0Jh", "groundedness_result": "3", "coherence_result": "4", "relevance_result": "4"},

gen_ai.response.id instead of id

File: src/api/evaluators/custom_evals/eval_combined.py (hunk 1 above)
Comment (Abdulk084) [reply src/api/evaluators/custom_evals/eval_combined.py]: fixed

File: src/api/evaluators/custom_evals/eval_combined.py (hunk 3 above)
Comment (Abdulk084) [reply src/api/evaluators/custom_evals/eval_combined.py:33]: fixed.

File: src/api/evaluators/custom_evals/eval_combined.py (hunk 2 above)
Comment (Abdulk084) [reply src/api/evaluators/custom_evals/eval_combined.py]: fixed.

**Comment by Abdulk084** [comment]
> [{"gen_ai.response.id": "chatcmpl-A8ibwkWtlnlbx4z5H17eGDeKQn0Jh", "groundedness_result": "3", "coherence_result": "4", "relevance_result": "4"},
> 
> gen_ai.response.id instead of id

Its been fixed now

`[{"gen_ai.response.id": "chatcmpl-A8ibwkWtlnlbx4z5H17eGDeKQn0Jh", "groundedness_result": "3"}, {"gen_ai.response.id": "chatcmpl-A8ibaHVRUWTajpnIibekjhKio0CZt", "groundedness_result": "The actual task input does not provide a context that can be used to evaluate the truthfulness of the answer provided. The answer seems to be a product recommendation for a specific brand of hiking shoes, but without a context that contains information about the TrailWalker Hiking Shoes or their features, it is not possible to determine whether the answer is true or false.\n\nTherefore, the correct evaluation score for the actual task input is:\n\n3"}]`

File: src/api/evaluators/custom_evals/eval_combined.py (hunk 4 above)
Comment (Abdulk084) [reply src/api/evaluators/custom_evals/eval_combined.py]: refactor done

**Review by drewby — APPROVED** [review]
Awesome work. Thanks for refactoring.

**Pull request merged**